File: backend/Testing/Browser/Page/AdminDashboardPage.cs
using Microsoft.Playwright;

namespace Testing.Browser.Page;

public class AdminDashboardPage : AuthenticatedBasePage<AdminDashboardPage>
{
    private ILocator ProjectFilterBarInput => Page.Locator(".filter-bar").Nth(0).GetByRole(AriaRole.Textbox);

    public AdminDashboardPage(IPage page)
    : base(page, "/admin", page.Locator(".breadcrumbs :text('Admin Dashboard')"))
    {
    }

    public async Task<ProjectPage> OpenProject(string projectName, string projectCode)
    {
        await ClickProject(projectName);
        return await new ProjectPage(Page, projectName, projectCode).WaitFor();
    }

    public async Task ClickProject(string projectName)
    {
        await ProjectFilterBarInput.FillAsync(projectName); // make sure the project is visible
        var projectTable = Page.Locator("table").Nth(0);
        await projectTable.GetByRole(AriaRole.Link, new() { Name = projectName, Exact = true }).ClickAsync();
    }
}

File: frontend/tests/pages/adminDashboardPage.ts
import type { Locator, Page } from '@playwright/test';

import { AuthenticatedBasePage } from './authenticatedBasePage';
import { CreateProjectPage } from './createProjectPage';
import { ProjectPage } from './projectPage';

export class AdminDashboardPage extends AuthenticatedBasePage {
  get projectFilterBarInput(): Locator { return this.page.locator('.filter-bar').nth(0).getByRole('textbox'); }

  constructor(page: Page) {
    super(page, page.locator(`.breadcrumbs :text('Admin Dashboard')`), `/admin`);
  }

  async openProject(projectName: string, projectCode: string): Promise<ProjectPage> {
    await this.clickProject(projectName);
    return await new ProjectPage(this.page, projectName, projectCode).waitFor();
  }

  async clickProject(projectName: string): Promise<void> {
    await this.projectFilterBarInput.fill(projectName); // make sure project is visible
    const table = this.page.locator('table').nth(0);
    return table.getByRole('link', {name: projectName, exact: true}).click();
  }

  async clickCreateProject(): Promise<CreateProjectPage> {
    await this.page.getByRole('link', {name: 'Create Project', exact: true}).click();
    return new CreateProjectPage(this.page).waitFor();
  }
}
